File: Project-Selection.java
import java.io.*;
import java.util.*;

// import java.util.ArrayDeque;
// import java.util.ArrayList;
// import java.util.StringTokenizer;

class FlowResults {
  ArrayList<ArrayList<Integer>> graph;
  Long maxFlow;
  ArrayList<Long> resCap;
  ArrayList<Integer> dest;

  public FlowResults() {
  }
}

class Arc {
  int u;
  int v;
  long c;

  public Arc(int u, int v, long c) {
    this.u = u;
    this.v = v;
    this.c = c;
  }
}

class NamedTuple {
  int curr;
  long bottleneck;
  int bottleneckStart;

  public NamedTuple(int curr, long bottleneck, int bottleneckStart) {
    this.curr = curr;
    this.bottleneck = bottleneck;
    this.bottleneckStart = bottleneckStart;
  }
}

class FlowNetwork {
  int m;
  int n;
  int s;
  int t;
  ArrayList<Long> cap;
  ArrayList<Integer> dest;
  ArrayList<ArrayList<Integer>> graph;

  public FlowNetwork(int m, int n, int s, int t, ArrayList<Arc> arcs) {
    this.m = m;
    this.n = n;
    this.s = s;
    this.t = t;

    cap = new ArrayList<Long>(2 * m);
    for (long i = 0; i < 2 * m; i++) {
      cap.add(0L);
    }

    dest = new ArrayList<Integer>(2 * m);

    graph = new ArrayList<>(n);
    for (long i = 0; i < n; i++) {
      graph.add(new ArrayList<>());
    }

    int edgeIdx = 0;
    for (Arc arc : arcs) {
      cap.set(edgeIdx, arc.c);
      graph.get(arc.u).add(edgeIdx++);
      graph.get(arc.v).add(edgeIdx++);
      dest.add(arc.v);
      dest.add(arc.u);
    }
  }

  private void computeDistances(ArrayList<Integer> distances, ArrayList<Integer> q, ArrayList<Long> resCap) {
    for (int i = 0; i < n; i++) {
      distances.set(i, -1);
    }
    distances.set(s, 0);

    int qs = 0;
    int qe = 1;
    q.set(0, s);

    while (qs < qe && distances.get(t) == -1) {
      int v = q.get(qs);
      qs++;
      if (distances.get(t) != -1 && distances.get(v) >= distances.get(t))
        break;

      for (int idx : graph.get(v)) {
        int w = dest.get(idx);
        if (resCap.get(idx) > 0 && distances.get(w) == -1) {
          distances.set(w, distances.get(v) + 1);
          q.set(qe, w);
          qe++;
        }
      }
    }
  }

  private long dinicAugment(ArrayList<Long> resCap, ArrayList<Integer> distances, ArrayList<Integer> ptrs) {
    long sSize = graph.get(s).size();
    long ans = 0;
    int curr = s;
    int bottleneckStart = -1;
    long bottleneck = Long.MAX_VALUE;

    ArrayDeque<NamedTuple> vertices = new ArrayDeque<>();

    if (ptrs.get(s) < sSize) {
      int idx = graph.get(s).get(ptrs.get(s));
      while (ptrs.get(s) < sSize && (distances.get(dest.get(idx)) != distances.get(s) + 1 || resCap.get(idx) == 0)) {
        ptrs.set(s, ptrs.get(s) + 1);
        if (ptrs.get(s) != graph.get(s).size())
          idx = graph.get(s).get(ptrs.get(s));
      }
    }

    while (ptrs.get(s) < sSize) {
      while (curr != t && ptrs.get(curr) < graph.get(curr).size()) {
        vertices.addFirst(new NamedTuple(curr, bottleneck, bottleneckStart));
        int idx = graph.get(curr).get(ptrs.get(curr));
        if (resCap.get(idx) < bottleneck) {
          bottleneck = resCap.get(idx);
          bottleneckStart = curr;
        }
        int next = dest.get(idx);

        if (ptrs.get(next) < graph.get(next).size()) {
          int idx2 = graph.get(next).get(ptrs.get(next));
          while (ptrs.get(next) < graph.get(next).size() &&
              (distances.get(dest.get(idx2)) != distances.get(next) + 1 || resCap.get(idx2) == 0)) {
            ptrs.set(next, ptrs.get(next) + 1);
            if (ptrs.get(next) != graph.get(next).size())
              idx2 = graph.get(next).get(ptrs.get(next));
          }
        }
        curr = next;
      }

      if (curr == t) {
        int curr2;
        while (!vertices.isEmpty()) {
          NamedTuple tup = vertices.removeFirst();
          curr2 = tup.curr;
          int idx = graph.get(curr2).get(ptrs.get(curr2));
          int revIdx = (cap.get(idx) == 0) ? idx - 1 : idx + 1;
          resCap.set(idx, resCap.get(idx) - bottleneck);
          resCap.set(revIdx, resCap.get(revIdx) + bottleneck);

          while (ptrs.get(curr2) < graph.get(curr2).size() &&
              (distances.get(dest.get(idx)) != distances.get(curr2) + 1 || resCap.get(idx) == 0)) {
            ptrs.set(curr2, ptrs.get(curr2) + 1);
            if (ptrs.get(curr2) != graph.get(curr2).size())
              idx = graph.get(curr2).get(ptrs.get(curr2));
          }
        }
        ans += bottleneck;

        bottleneck = Long.MAX_VALUE;
        bottleneckStart = -1;
        curr = s;
      } else {
        int pred = vertices.getFirst().curr;
        while (ptrs.get(curr) == graph.get(curr).size() && !vertices.isEmpty()) {
          NamedTuple tup = vertices.removeFirst();
          curr = tup.curr;
          bottleneck = tup.bottleneck;
          bottleneckStart = tup.bottleneckStart;
          ptrs.set(curr, ptrs.get(curr) + 1);

          if (ptrs.get(curr) < graph.get(curr).size()) {
            int idx = graph.get(curr).get(ptrs.get(curr));
            while (ptrs.get(curr) < graph.get(curr).size() &&
                (distances.get(dest.get(idx)) != distances.get(curr) + 1 || resCap.get(idx) == 0)) {
              ptrs.set(curr, ptrs.get(curr) + 1);
              if (ptrs.get(curr) != graph.get(curr).size())
                idx = graph.get(curr).get(ptrs.get(curr));
            }
          }
        }

        if (vertices.isEmpty()) {
          bottleneck = Long.MAX_VALUE;
          bottleneckStart = -1;
          curr = s;
        }
      }
    }
    return ans;
  }

  FlowResults dinic() {
    long ans = 0;
    ArrayList<Long> resCap = new ArrayList<>(cap);
    ArrayList<Integer> q = new ArrayList<>(2 * n);
    for (long i = 0; i < 2 * n; i++) {
      q.add(0);
    }

    ArrayList<Integer> distances = new ArrayList<>(n);
    for (long i = 0; i < n; i++) {
      distances.add(-1);
    }

    ArrayList<Integer> ptrs = new ArrayList<>(n);
    for (long i = 0; i < n; i++) {
      ptrs.add(0);
    }

    computeDistances(distances, q, resCap);

    while (distances.get(t) != -1) {
      ans += dinicAugment(resCap, distances, ptrs);
      computeDistances(distances, q, resCap);
      for (int i = 0; i < n; i++) {
        ptrs.set(i, 0);
      }
    }

    FlowResults fr = new FlowResults();
    fr.dest = dest;
    fr.maxFlow = ans;
    fr.graph = graph;
    fr.resCap = resCap;
    return fr;
  }
}

class Main {

  public static int N, M, n, m;

  // public static ArrayList<Integer> bfs(FlowResults fR) {

  // return null;
  // }

  public static ArrayList<Integer> bfs(FlowResults fR) {
    // Initialize the array to keep track of visited nodes
    boolean[] visited = new boolean[N];

    // Initialize the queue for BFS and add the starting node (0)
    Queue<Integer> queue = new LinkedList<>();
    queue.add(0);
    visited[0] = true;

    // Initialize the result ArrayList
    ArrayList<Integer> ans = new ArrayList<>();

    while (!queue.isEmpty()) {
      int current = queue.poll();

      // Check all adjacent edge indices of the current node
      for (int edgeIndex : fR.graph.get(current)) {
        int nextNode = fR.dest.get(edgeIndex);
        long edgeWeight = fR.resCap.get(edgeIndex);

        // Check if the edge has weight > 0 and nextNode is not visited
        if (edgeWeight > 0 && !visited[nextNode]) {
          queue.add(nextNode);
          visited[nextNode] = true;

          // Check if the index of the reachable node is between 1 and n
          if (nextNode >= 1 && nextNode <= n) {
            ans.add(nextNode);
          }
        }
      }
    }

    return ans;
  }

  public static void main(String[] args) throws IOException {
    final int inf = Integer.MAX_VALUE;

    // total profit without paying cleaning cost
    long w = 0L;

    // Step 1: Read Inputs and Build Arcs
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    String[] line1 = br.readLine().split(" ");
    assert line1.length == 3;
    n = Integer.parseInt(line1[0]); // n is the number of vertices
    m = Integer.parseInt(line1[1]); // m is the number of edges
    int p = Integer.parseInt(line1[2]); // p is the multiplier
    N = 2 + n + m;
    M = 3 * m + n;

    ArrayList<Arc> arcs = new ArrayList<>(); // arcs store all edges of the input graph G'

    for (int i = 1; i <= n; i++) {
      int s = Integer.parseInt(br.readLine()) * p;
      w += s;
      arcs.add(new Arc(0, i, s));
    }

    for (int i = 1; i <= m; i++) {
      String[] uvc = br.readLine().split(" ");
      // int u = Integer.parseInt(uvc[0]);
      // int v = Integer.parseInt(uvc[1]);
      // int c = Integer.parseInt(uvc[2]);
      arcs.add(new Arc(Integer.parseInt(uvc[0]), n + i, inf)); // edge from u to uv
      arcs.add(new Arc(Integer.parseInt(uvc[1]), n + i, inf)); // edge from v to uv
      arcs.add(new Arc(n + i, n + m + 1, Integer.parseInt(uvc[2]))); // edge from uv to t
    }

    br.close();

    // Step 2: Create the Flow Network
    final int source = 0;
    final int sink = m + n + 1;

    FlowNetwork fN = new FlowNetwork(arcs.size(), 2 + m + n, source, sink, arcs);

    // Step 3: Use the Dinic algorithm to find the max flow
    FlowResults fR = fN.dinic();

    // Step 4: Use bfs to get all reachable nodes from s
    ArrayList<Integer> output = bfs(fR);

    // Step 5: Output the result
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    bw.write(Integer.toString((int) (w - fR.maxFlow)));
    bw.write(" ");
    bw.write(Integer.toString(output.size()));
    // bw.newLine();
    bw.write("\n");

    for (int i = 0; i < output.size(); i++) {
      bw.write(Integer.toString(output.get(i)));
      if (i != output.size() - 1)
        bw.write(" ");
    }
    bw.write("\n");

    bw.close();
  }
}

File: Project-Selection.cpp
#ifndef ALGO_CPP_FLOWNETWORK_H
#define ALGO_CPP_FLOWNETWORK_H

#include <cstdint>
#include <limits>
#include <vector>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>
#include <iostream>
#include <limits.h>
#include <stdio.h>

namespace flow_network
{
  class Arc
  {
  public:
    uint64_t u;
    uint64_t v;
    uint64_t c;

    Arc(uint64_t u, uint64_t v, uint64_t c);
    Arc();
  };

  class FlowResults
  {
  public:
    FlowResults(std::vector<std::vector<uint64_t>> graph, std::vector<uint64_t> dest);
    uint64_t max_flow;
    std::vector<uint64_t> res_cap;
    const std::vector<std::vector<uint64_t>> graph;
    const std::vector<uint64_t> dest;
  };

  class FlowNetwork
  {

  public:
    FlowNetwork(uint64_t m, uint64_t n, uint64_t s, uint64_t t, const std::vector<Arc> &arcs);
    FlowResults dinic();

  private:
    std::vector<std::vector<uint64_t>> graph;
    uint64_t s;
    uint64_t t;
    uint64_t m;
    uint64_t n;
    std::vector<uint64_t> cap;
    std::vector<uint64_t> dest;
  };

}

#endif // ALGO_CPP_FLOWNETWORK_H

flow_network::Arc::Arc(uint64_t u, uint64_t v, uint64_t c) : u(u), v(v), c(c){};

flow_network::Arc::Arc() = default;

flow_network::FlowResults::FlowResults(std::vector<std::vector<uint64_t>> graph, std::vector<uint64_t> dest) : graph(
                                                                                                                   std::move(graph)),
                                                                                                               dest(std::move(dest)) {}

flow_network::FlowNetwork::FlowNetwork(uint64_t m, uint64_t n, uint64_t s, uint64_t t, const std::vector<Arc> &arcs)
    : m(m), n(n), s(s), t(t)
{
  cap = std::vector<uint64_t>(2 * m, 0);
  graph = std::vector<std::vector<uint64_t>>(n);
  uint64_t edge_idx = 0;
  for (auto &arc : arcs)
  {
    cap[edge_idx] = arc.c;
    graph[arc.u].push_back(edge_idx++);
    graph[arc.v].push_back(edge_idx++);
    dest.push_back(arc.v);
    dest.push_back(arc.u);
  }
}

#include <vector>
void compute_distances_faster(std::vector<uint64_t> &distances, std::vector<uint64_t> &q, const std::vector<uint64_t> &rescap,
                              const std::vector<std::vector<uint64_t>> &graph, const std::vector<uint64_t> &dest, uint64_t s,
                              uint64_t t)
{
  std::fill(distances.begin(), distances.end(), -1);
  uint64_t n = graph.size();

  distances[s] = 0;
  uint64_t qs = 0;
  uint64_t qe = 1;
  q[0] = s;

  while (qs < qe && distances[t] == -1)
  {
    uint64_t v = q[qs];
    qs++;
    if (distances[t] != -1 && distances[v] >= distances[t])
      break;
    for (uint64_t idx : graph[v])
    {
      uint64_t w = dest[idx];
      if (rescap[idx] > 0 && distances[w] == uint64_t(-1))
      {
        distances[w] = distances[v] + 1;
        q[qe] = w;
        qe++;
      }
    }
  }
}

// use some tricks to augment by a blocking flow in O(mn) time
uint64_t dinic_augment(std::vector<uint64_t> &rescap, const std::vector<uint64_t> &cap, std::vector<uint64_t> &ptrs,
                       const std::vector<uint64_t> &distances, const std::vector<std::vector<uint64_t>> &graph,
                       const std::vector<uint64_t> &dest, uint64_t s, uint64_t t)
{

  const uint64_t s_size = graph[s].size();

  uint64_t ans = 0;
  // each record [u, bottleneck_cap, bottleneck_start] means that there's a path from s to u with bottleneck edge leaving bottleneck_start of capacity bottleneck_cap
  std::stack<std::tuple<uint64_t, uint64_t, uint64_t>> vertices;

  // state of search
  uint64_t curr;
  uint64_t bottleneck;
  uint64_t bottleneck_start;

  // fix ptr for s, init first search
  {
    uint64_t idx = graph[s][ptrs[s]];
    while (ptrs[s] < graph[s].size() && (distances[dest[idx]] != distances[s] + 1 || rescap[idx] == 0))
    {
      ptrs[s]++;
      idx = graph[s][ptrs[s]];
    }

    bottleneck = std::numeric_limits<uint64_t>::max();
    bottleneck_start = -1;
    curr = s;
  }

  // while there's an edge leaving s in level graph
  while (ptrs[s] < s_size)
  {
    // basic invariant that I spammed everywhere

    // stop whenever we find t or a dead end (and don't put them on the stack, instead store using 3 state vars)
    while (curr != t && ptrs[curr] < graph[curr].size())
    {
      vertices.emplace(curr, bottleneck, bottleneck_start);

      uint64_t idx = graph[curr][ptrs[curr]];

      if (rescap[idx] < bottleneck)
      {
        bottleneck = rescap[idx];
        bottleneck_start = curr;
      }
      uint64_t next = dest[idx];

      // fix ptr for next (to check if it's a dead end)
      {
        uint64_t idx = graph[next][ptrs[next]];
        while (ptrs[next] < graph[next].size() &&
               (distances[dest[idx]] != distances[next] + 1 || rescap[idx] == 0))
        {
          ptrs[next]++;
          idx = graph[next][ptrs[next]];
        }
      }

      curr = next;

      {
        auto tup = vertices.top();
        uint64_t curr2 = std::get<0>(tup);
        uint64_t idx = graph[curr2][ptrs[curr2]];
        uint64_t revIdx = (cap[idx] == 0) ? idx - 1 : idx + 1;
      }
    }

    if (curr == t)
    {
      //            fprintf(stderr, "augmenting\n");
      uint64_t curr2;
      while (!vertices.empty())
      {
        auto tup = vertices.top();
        curr2 = std::get<0>(tup);

        // graph formation invariant: forward edges precede their neighboring backward edges
        uint64_t idx = graph[curr2][ptrs[curr2]];
        uint64_t revIdx = (cap[idx] == 0) ? idx - 1 : idx + 1;

        rescap[idx] -= bottleneck;
        rescap[revIdx] += bottleneck;

        while (ptrs[curr2] < graph[curr2].size() &&
               (distances[dest[idx]] != distances[curr2] + 1 || rescap[idx] == 0))
        {
          ptrs[curr2]++;
          idx = graph[curr2][ptrs[curr2]];
        }
        vertices.pop();
      }
      ans += bottleneck;

      // do another search
      bottleneck = std::numeric_limits<uint64_t>::max();
      bottleneck_start = -1;
      curr = s;
    }
    else
    {
      // delete edges leading to dead-end vertices

      uint64_t pred = std::get<0>(vertices.top()); // direct predecessor of node with outdegree 0

      while (ptrs[curr] == graph[curr].size() && !vertices.empty())
      { // curr has out-degree 0 in level graph

        auto tup = vertices.top();
        curr = std::get<0>(tup);
        bottleneck = std::get<1>(tup);
        bottleneck_start = std::get<2>(tup);

        vertices.pop();

        ptrs[curr]++; // discard the edge that led to the dead end

        { // discard other edges
          uint64_t idx = graph[curr][ptrs[curr]];
          while (ptrs[curr] < graph[curr].size() &&
                 (distances[dest[idx]] != distances[curr] + 1 || rescap[idx] == 0))
          {
            ptrs[curr]++;
            idx = graph[curr][ptrs[curr]];
          }
        }
      }

      // check that some pointer went up but not uselessly

      if (!vertices.empty())
      {
        bottleneck = bottleneck;
        bottleneck_start = bottleneck_start;
        curr = curr;
      }
      else
      {
        bottleneck = std::numeric_limits<uint64_t>::max();
        bottleneck_start = -1;
        curr = s;
      }
    }
  }

  return ans;
}

flow_network::FlowResults flow_network::FlowNetwork::dinic()
{
  uint64_t ans = 0;
  std::vector<uint64_t> rescap(cap);
  std::vector<uint64_t> q(2 * n);

  // result of BFS (to avoid actually manifesting the edges of the level graph)
  std::vector<uint64_t> distances(n, -1);
  std::vector<uint64_t> ptrs(n, 0); // ptr[u] == k  <==>  the first k neighbors of u are not in the level graph
                                    //    compute_distances(distances, rescap, graph, dest, s, t);
  compute_distances_faster(distances, q, rescap, graph, dest, s, t);

  while (distances[t] != uint64_t(-1))
  {
    //        fprintf(stderr, "t dist: %lld\n", distances[t]);
    ans += dinic_augment(rescap, cap, ptrs, distances, graph, dest, s, t);
    compute_distances_faster(distances, q, rescap, graph, dest, s, t);

    std::fill(ptrs.begin(), ptrs.end(), 0); // reset ptrs
  }

  FlowResults fr(graph, dest);
  fr.res_cap = rescap;
  fr.max_flow = ans;

  return fr;
}

using namespace std;

static int N, M, n, m;

vector<int> bfs(flow_network::FlowResults fR)
{
  // Initialize the array to keep track of visited nodes
  vector<bool> visited(N, false);

  // Initialize the queue for BFS and add the starting node (0)
  queue<int> queue;
  queue.push(0);
  visited[0] = true;

  // Initialize the result vector
  vector<int> ans;

  while (!queue.empty())
  {
    int current = queue.front();
    queue.pop();

    // Check all adjacent edge indices of the current node
    for (int edgeIndex : fR.graph[current])
    {
      int nextNode = fR.dest[edgeIndex];
      long edgeWeight = fR.res_cap[edgeIndex];

      // Check if the edge has weight > 0 and nextNode is not visited
      if (edgeWeight > 0 && !visited[nextNode])
      {
        queue.push(nextNode);
        visited[nextNode] = true;

        // Check if the index of the reachable node is between 1 and n
        if (nextNode >= 1 && nextNode <= n)
        {
          ans.push_back(nextNode);
        }
      }
    }
  }

  return ans;
}

int main()
{
  const int inf = INT_MAX;

  // total profit without paying cleaning cost
  long w = 0L;

  // Step 1: Read Inputs and Build Arcs
  int p;
  scanf("%d %d %d", &n, &m, &p); // n is the number of vertices, m is the number of edges, p is the multiplier
  N = 2 + n + m;
  M = 3 * m + n;

  vector<flow_network::Arc> arcs; // arcs store all edges of the input graph G'

  for (int i = 1; i <= n; i++)
  {
    int s;
    scanf("%d", &s);
    s *= p;
    w += s;
    arcs.push_back(flow_network::Arc(0, i, s));
  }

  for (int i = 1; i <= m; i++)
  {
    int u, v, c;
    scanf("%d %d %d", &u, &v, &c);
    arcs.push_back(flow_network::Arc(u, n + i, inf));       // edge from u to uv
    arcs.push_back(flow_network::Arc(v, n + i, inf));       // edge from v to uv
    arcs.push_back(flow_network::Arc(n + i, n + m + 1, c)); // edge from uv to t
  }

  // Step 2: Create the Flow Network
  const int source = 0;
  const int sink = m + n + 1;

  flow_network::FlowNetwork fN(arcs.size(), 2 + m + n, source, sink, arcs);

  // Step 3: Use the Dinic algorithm to find the max flow
  flow_network::FlowResults fR = fN.dinic();

  // Step 4: Use bfs to get all reachable nodes from s
  vector<int> output = bfs(fR);

  // Step 5: Output the result
  printf("%d %lu\n", (int)(w - fR.max_flow), output.size());

  for (size_t i = 0; i < output.size(); ++i)
  {
    printf("%d", output[i]);
    if (i != output.size() - 1)
      printf(" ");
  }

  printf("\n");

  return 0;
}
